pkg/cmd/search/shared/searcher.go

```diff
@@ -0,0 +1,178 @@
+package shared
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"math"
+	"net/http"
+	"net/url"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/cli/cli/v2/pkg/search"
+)
+
+type searcher struct {
+	client *http.Client
+	host   string
+}
+
+func NewSearcher(client *http.Client, host string) *searcher {
+	return &searcher{
+		client: client,
+		host:   host,
+	}
+}
+
+func (s *searcher) Search(query search.Query) (search.Result, error) {
+	result := search.Result{}
+	path := fmt.Sprintf("https://api.%s/search/%s", s.host, query.Kind)
+	queryString := url.Values{}
+	q := strings.Builder{}
+	quotedKeywords := quoteKeywords(query.Keywords)
+	q.WriteString(strings.Join(quotedKeywords, " "))
+	for k, v := range query.Qualifiers.ListSet() {
+		v = quoteQualifier(v)
+		q.WriteString(fmt.Sprintf(" %s:%s", k, v))
+	}
+	queryString.Set("q", q.String())
+	if query.Order.IsSet() {
+		queryString.Set(query.Order.Key(), query.Order.String())
+	}
+	if query.Sort.IsSet() {
+		queryString.Set(query.Sort.Key(), query.Sort.String())
+	}
+	pages := math.Ceil((float64(query.Limit) / float64(100)))
+	for i := 1; i <= int(pages); i++ {
+		queryString.Set("page", strconv.Itoa(i))
+		remaining := query.Limit - (i * 100)
+		if remaining > 100 {
+			queryString.Set("per_page", "100")
+		} else if remaining <= 0 {
+			queryString.Set("per_page", strconv.Itoa(query.Limit))
+		} else {
+			queryString.Set("per_page", strconv.Itoa(remaining))
+		}
+		url := fmt.Sprintf("%s?%s", path, queryString.Encode())
+		req, err := http.NewRequest("GET", url, nil)
+		if err != nil {
+			return result, err
+		}
+		req.Header.Set("Content-Type", "application/json; charset=utf-8")
+		req.Header.Set("Accept", "application/vnd.github.v3+json")
+		resp, err := s.client.Do(req)
+		if err != nil {
+			return result, err
+		}
+		defer resp.Body.Close()
+		success := resp.StatusCode >= 200 && resp.StatusCode < 300
+		if !success {
+			return result, handleHTTPError(resp)
+		}
+		b, err := io.ReadAll(resp.Body)
+		if err != nil {
+			return result, err
+		}
+		pageResult := search.Result{}
+		err = json.Unmarshal(b, &pageResult)
```

Comment:
Protip: using a `json.Decoder` you can unmarshal `resp.Body` (or any Reader) directly into the result struct without having to use an intermediate `b` bytes slice.